plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MavenPathParser.java

```diff
@@ -106,23 +108,41 @@ else if (str.endsWith("." + HashType.MD5.getExt())) {
           tail = str.substring(nTailPos);
         }
         else {
-          StringBuilder snapshotBuildNumber = new StringBuilder(version);
-          snapshotBuildNumber
-              .append(str.substring(vSnapshotStart + version.length(), vSnapshotStart + version.length() + 8));
+          final StringBuilder snapshotTimestampedVersion = new StringBuilder(version);
+          snapshotTimestampedVersion.append(
+              str.substring(vSnapshotStart + version.length(), vSnapshotStart + version.length() + 7)
+          );
+
+          try {
+            timestamp = Maven2Format.METADATA_DOTTED_TIMESTAMP.parseDateTime(
+                snapshotTimestampedVersion.toString()).getMillis();
+          }
+          catch (IllegalArgumentException e) {
+            // skip it
+          }
+
+          // add the dash between timestamp and buildNo
+          snapshotTimestampedVersion.append('-');
 
-          int buildNumberCount = 0;
-          int buildNumberPos = vSnapshotStart + snapshotBuildNumber.length();
+          int buildNumberPos = vSnapshotStart + snapshotTimestampedVersion.length();
+          final StringBuilder bnr = new StringBuilder();
           while (str.charAt(buildNumberPos) >= '0' && str.charAt(buildNumberPos) <= '9') {
-            snapshotBuildNumber.append(str.charAt(buildNumberPos));
+            snapshotTimestampedVersion.append(str.charAt(buildNumberPos));
+            bnr.append(str.charAt(buildNumberPos));
             buildNumberPos++;
-            buildNumberCount++;
           }
-          if (buildNumberCount == 0) {
+          if (bnr.length() == 0) {
             return null;
           }
+          try {
+            buildNumber = Integer.parseInt(bnr.toString());
```

Comment:
since the above loop is just accumulating digits it is possible to accumulate the build number in the loop rather than build a numeric string and then convert it